### ISSUE TITLE
Sensors data availability ?

### ISSUE THREAD

**Issue by qgetomat** [JH-Soft-Technology/InelsForHass#15]
Dears.  1st big thanks for the project. Quite well working.
Is there any chances to get the sensors data   Like temperature or humidity?

many thanks

## Configuration

```yaml

Add your logs here.

```

## Debug log

```text

Add your logs here.

```

**Comment by jhoralek** [comment]
Hi there,

What kind of temperature and humidity sensor you mean? 

Has Inels some kind of these sensors on BUS protocol? 

Kind regards.

JH.

**Comment by ds141414** [comment]
Hello! 
Yes Inels has two ways of reporting temperature.  The first is with the built in temperature sensor of the wall switches WS-20 and WS-40 and by adding an external temperature sensor  in the analogue inputs of the switches.  From what I understatand any Inels device that has an analogue input we use a temperature sensor.  The central unit can be used as an interface for it.

This is how I read them in Home Assistant

  - platform: tcp
    name: "Outdoor-Roof Temp"
    host: 192.168.1.xx
    port: 1111
    timeout: 2
    payload: "GET,0x01050041\r\n"
    value_template: "{{ value.split(',')[2] | int(base=10)/100 }}"
    unit_of_measurement: "°C"


I will post screenshots and photos of the actual switches and sensors later on today.

Kind regards,
DS

**Comment by jhoralek** [comment]
It is interesting. 

This integration is communicating with CU unit through the connection server and there is no way to touch wsb switches at all.

So I'm a bit interesting the way you are reaching the unit.

What version number of the CU unit, you have?

They have many firmware versions even on the same CU type, so it can be very difficult to get data, at the same format.

I'm going to try your code on my CU unit.

I'll be in touch with you soon and share my results.


Kind regards

JH

**Comment by ds141414** [comment]
Hello,

Yes it was intersting finding things out about how to interface with the CU.  I managed to communicate directly through the telnet connection protocol.  You have to use IDM to set that up.  Once you telnet into the CU you can send commands

My CU firmware version is, there is an upgrade but I am hessitant to upgrade it in case it breaks anything.
![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/5569eed6-f9e9-48cf-986b-51dd34835ac8)

You have to enable Third Party Communicatin through IDM

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/60458bc9-4e02-4dab-befa-c37539a8a227)

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/1070d983-50bd-4acd-87fd-5bbcb7184c12)


This is what comes up on the screen through Putty

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/a06b07d1-6e00-4f20-b25f-979b96f7678c)

You can get the code and name for each device from the Connection server 
![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/1f89408f-228a-42e5-a940-e6bfc0d91f4e)




![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/9b3e2007-60c6-4234-bc9f-86706f2cf0bc)

This is I have it setup in Connection server

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/d6d4f81a-a01b-40f3-ae4d-f89d688f820b)

This is how I setup a new sensor.  It will later on appear on my Inels phone App

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/97ff2634-cc82-46f4-83fe-46d8acd4c885)

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/2d767c93-12ee-4266-87b4-ead07833de69)



You could probably use this read the temperature from the connection server and create a sensor in Home Assistant.

The tcp sensors in Home Assistant work pretty reliably so far.  I use it to keep track of my hot water boiler on the roof and it has been pretty solid.  I created a climate entity in HA and control on/off of hot water from there.  Paired that with is an NSPanel it is sweet :) 

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/a33ce18a-1dcb-4839-a889-27957355f279)

![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/d0faf1d1-8ca1-4aa8-b3a5-dd88f830f24a)


I hope this will be of some sort help to anyone looking to integrate Inels CU temperature sensor to HA.

Kind regards,
DS

PS
This is the temperature sensor on the wall switch, I have this in the garage so I left it without the cover.
![image](https://github.com/JH-Soft-Technology/InelsForHass/assets/61230688/0a5716af-e5f1-4dab-b0b4-144df1447dec)

**Comment by ds141414** [comment]
Temperature for boiler and NS panel are different because I took the pictures at different times :)

**Comment by stomko11** [comment]
@ds141414 can you shed some more light on using tcp sensor? Friend of mine has inels. not sure which version of CU. But he is using thermometers in rooms, and this is currently not supported by this integration. I am curious if there is any way to get around it - perhaps by querying CU to get raw data that I could parse later. Any help would be appreciated

**Comment by ds141414** [comment]
Hi! No problem,  I will get back to you this afternoon when I am home and gave access to CU.

**Comment by ds141414** [comment]
Hello again!  
Have you seen my post [here](https://github.com/JH-Soft-Technology/InelsForHass/issues/15#issuecomment-1881535778), above actually.  It explains everything how it is done.  I use Home Assistant to create the TCP sensor.

Is there anything specific that you need cleared up? 

Happy to help!

**Comment by stomko11** [comment]
Not sure how I missed it. But I looked and friend's CU is CU3-01M, fw version 02.8C.00 and his idm3 central unit software is version 3.3.27. i checked on inels site, they have only FWs and idm3 for cu3-07m and above. Perhaps you have newer version somewhere? Otherwise I would need to reach out to inels for latest idm3/fw

**Comment by ds141414** [comment]
Good morning,

I have the same CU, CU3-01M.  This functionality should be available with the 3.3.27 version of IDM but I provide below two links for 3.4.18 and 3.4.19.  The last  one includes an MQTT client.  I haven't been able to make it work as yet, you may want to try it out.



![image](https://github.com/user-attachments/assets/44ce71c7-5272-4f11-9fae-f5719e6927fe)



[3.4.18]( https://mydrive.seven7web.com/s/EijrXdgfK2jRkjo)
[3.4.19](https://mydrive.seven7web.com/s/M6EGCe2WwF86Yqr)

![image](https://github.com/user-attachments/assets/ab00ac44-e8d4-40c3-a335-d19cebb92885)

Let me know how you get along!

Have a nice day!

**Comment by stomko11** [comment]
well, I also received response from Inels - they said MQTT is supported, but in case of CU3-01M and 02-M, it is only for system bits, not for inels modules. That is available only for CU3-07M and above.
For TCP sensor, I will give it a try and let you know. i did notice that support is there, output is bit different. So I may try updating it first.

**Comment by ds141414** [comment]
Yes you are right it is for system bits.  However system bits are very easy to work with as you can link them with any on/off device and copy the state.  The issue I faced with that was that after activating MQTT I would lose connection to the CU from IDM.  After a cold restart I would be able to connect back.  Since I am able to control everything with CS and HA integration I kind of put it on the back burner.

I am curious how you fare out

Let me know

**Comment by stomko11** [comment]
Are system bits only for on/off devices? Means these would not include temperature sensors? If yes, it is pointless for me to even try since i can get all other switches/lights through these integrations.
I am interested in temp sensors.
One thing I did not quite get (yet) - is the tcp sensor polling, or is it constantly listening for events? My understanding is that you need to send a payload to get response from CU (that is what payload is there for in definition), but then - how often is HA sending that query?
So far I have asked my friend to enable 3rd party communication, i can telnet into CU on port 1111 but if I just listen for the output in putty (not sending any payload), it just reports "NOP"

**Comment by stomko11** [comment]
Could you perhaps share example of device, how you query it as well as how is it listed in export?
For instance, I found following in export:
'heat_garage 0x02080008 0x02`
and
`heat_garage_Temperature Y B 2181103624 REAL PUB_INOUT`

I think my friend did not configure separator, so I have changed it (left it as space, but I have also tried with `GET,xxx` which actually returns timeout), to see raw value that is returned
```
- platform: tcp
  name: "Temp Test Garage"
  host: 192.168.x.x
  port: 1111
  timeout: 2
  payload: "GET 0x02080008\r\n"
  value_template: "{{ value }}"
```

But this is actually returning `GET 0x02080004 N`
Therefore I am not sure if I am doing something wrong, or perhaps older firmware acts differently

**Comment by ds141414** [comment]
Hello,

Did you try to interact directly with  the CU via the terminal? That's how I started .  I also set up the separator to be a comma, that way I get single string separated by commas and I pick the value after the second comma.  Have a look below.  

This is the sensor from my solar water heater.  The TCP sensor and below the raw data as it is returned from the CU.

Also have a look at this manual from Inels.  It is their manual for ASCII communication with the CU.

[idm3_tutorial17_ascii-communication-protocol.pdf](https://github.com/user-attachments/files/17330035/idm3_tutorial17_ascii-communication-protocol.pdf)


This is my CU setup, notice the comma as the separator.  

![image](https://github.com/user-attachments/assets/18d4571b-039c-4337-879e-641971fcf718)


Also use this this to download the imm file

**cu-ip-address/immfiles/export.imm**

Here is a sample of my file that includes the example below.  Also notice that the name of the device providing the temperature from the wall switches begins with something like: 

WSB3-40_AIN1-AIN2-Therm_017C15 0x01050008 0x00000000 °C

![image](https://github.com/user-attachments/assets/d70830e8-0ed3-49fc-af49-8d5376873576)


 - platform: tcp
    name: water_heater_temp
    host: 192.168.1.xx
    port: 1111
    timeout: 2
    payload: "GET,0x01050009\r\n"
    value_template: "{{ value.split(',')[2] | int(base=10)/100 }}"
    unit_of_measurement: "°C"


GET,0x01050009
GET,0x01050009,7994


This should give you a good idea!

**Comment by stomko11** [comment]
This helped a lot. One of issues that I was doing is that not all WSB3-40 were exposed. Instead, they were used internally in heating circuits, which was exported but I was unable to query that (which is expected).
Since there are multiple devices I would like to query, I tried to follow the guide you provided, that says user can query multiple parameters in one command:
```
NOTE: user can stick several requests together by separating each command using special
symbol “;”, e.g. GET 0x01020045 ; GET 0x01020046 ; GET 0x01020047, therefore third-party
application can make only one request instead of three.
```
Unfortunately, this is not working for me - comma separator is used, but even if I try it in this format:
```
GET,0x01020045 ; GET,0x01020046 ; GET,0x01020047
```
It does not return any value. Is this perhaps working for you? I tried that using putty. Querying one by one works fine. 
Is there peraps any downside to querying 30+ parameters, each in its own command?

Thank you.

**Comment by ds141414** [comment]
Hi! If you are getting results when you use it one by one then you are OK.  The Home Assistant sensor can only take one input per sensor.

Did you try to setup an HA sensor ?

I assumed you already had the temp variable marked for export and I did not put it in the guide

**Comment by stomko11** [comment]
Sensor works fine. But since I want to query multiple, I wanted to avoid setting multiple TCP sensors and instead, send one command with multiple GET requests, as instructed in manual, and then create sensor based on answer as described here in post #2 https://community.home-assistant.io/t/is-it-possible-to-consigure-one-tcp-sensor-request-and-multiple-sensors-from-the-same-one-response/285846/2
But in my case, CU just won't respond anything if I bundle multiple GET commands together.

**Comment by ds141414** [comment]
Hi! 

I was not aware of this option in HA.  I  too have a number of top sensors set up and I don't really like it as it is not a very efficient way to do it.

I am out now, I will try in the afternoon and let you know what happens 

Thanks for the info

**Comment by stomko11** [comment]
Thank you. It would be interesting to know if you can query CU via Putty in that way. I was unsuccessful. But I am also still on the older version of FW (02.8C.00).

**Comment by stomko11** [comment]
So I have tried adding around 35 sensors today and clearly CU does not like it. It resets connections (perhaps if too many are opened). I have reduced number to 18, 6 groups with 3 sensors with update interval 10s different from the other. That works, but it is certainly not neat as one cannot rely on the data being received (or not). So wondering if you managed to get multiple GET requests bundled together and sent to CU.

**Comment by ds141414** [comment]
Hello!
Crazy couple of days but I finally found some time to work on this.  Your suggestion of collecting everything is a single sensor really helped.  I set this up a long time ago, before creating the TCP sensor and finding this repo. I set up Telnet switches to control my lights.  Indeed the CU was not happy about that.  Sometime the switches would not initialize.  

So this is how I did it.  I based my solution on the post you included a couple of days ago. 

With this sensor I am getting all the data from CU.  I guess it is working OK with me when I send the request as a string

sensor:
  - platform: tcp
    name: inels_sensors
    host: 192.168.x.xx
    port: 1111
    timeout: 2
    payload: "GET,0x01080007;GET,0x01050009;GET,0x0105000A;GET,0x01050021;GET,0x0105002B;GET,0x01050036;GET,0x01050037;GET,0x0105003F;GET,0x01050029\r\n"
    value_template: "{{ value }}"

This stores all the data I get from the CU in the {{vallue}}  variable.  From there all I did was to crate a template sensor and I stacked them up like this

- template
   
  - sensor:
      - name: Inels Water Heater Temp
        unit_of_measurement: "°C"
        state: >
          **{% set state = states('sensor.inels_sensors')%}
          {{state.split(',')[4] | int(base=10)/100 }}**
      - name: Inels Roof Temp
        unit_of_measurement: "°C"
        state: >
          **{% set state = states('sensor.inels_sensors')%}
          {{state.split(',')[6] | int(base=10)/100 }}**

      - name: Room1 Temp
        unit_of_measurement: "°C"
        state: >
          **{% set state = states('sensor.inels_sensors')%}
          {{state.split(',')[6] | int(base=10)/100 }}**

      - name: Room2 Temp
        unit_of_measurement: "°C"
        state: >
          **{% set state = states('sensor.inels_sensors')%}
          {{state.split(',')[8] | int(base=10)/100 }}**
 etc

Notice how the subscript increases when I move to the next sensor.  Additionally you could use the templates from Helpers to create your sensor.  The downside to that is creating the sensors one by one.

This creates the sensors based on the data from the first sensor.  To make it work I restarted Home Assistant before creating the template sensors.  I did this so that the **sensor.inels_sensors**  will be created and I could reference it in the Template Generator to make sure everything was working as expected.

It sound weird that the CU is able to respond back with a single request and it does not with multiple.  I don't think this is a firmware issue because I have been running this for a very long time, way before I upgraded to the latest version.

Here are my CU settings
![image](https://github.com/user-attachments/assets/123a9bcd-8c70-4315-a2dc-92b5254e50af)

![image](https://github.com/user-attachments/assets/725e8f8e-67aa-4993-9361-562d03000809)

Also this is what I get by sending a single request to my CU

![image](https://github.com/user-attachments/assets/f306cf49-6842-4ef1-b71c-7bf277ef08bd)

**Comment by stomko11** [comment]
Thanks for response. I don't know honestly. What I do know is that friend of mine runs original firmware for around last 5 years. I guess he will attempt upgrade again to see if it makes any difference.
I have also reached out to Inels, hoping to get an answer, but the answer was that delimitere needs to be set to blank space, which is obviously incorrect since you use "," and it works.
They also advised that MQTT for system bits is available from 3.4.19 rev5 (your link was for rev3), here is link for [rev7](https://nextcloud.stomko.net/s/iN3xwdyRPeKGnaJ) - this is a test version

**Comment by stomko11** [comment]
To give an update - attempted to do upgrade to latest FW available in 3.4.19, querying multiple values worked there. However I ran into issues with connection server - that one somehow stopped working, as a result both inels app in phone as well as this HA intergration (since it connects to CS) stopped working. Rolled back to IDM 3.3.x used before - it worked, but was not straight forward as last time. FW got updated to original one couple of times, but then again after connecting it reported v00.00.00 and required CU upgrade. Anyway, we got past it somehow, but CS is still broken. 
I am a bit frustrated with this all "ease of use", complicated documentation, and unclear description of what version works with what FW, etc. Such as - you have IDM for basic project creation. That needed to be converted to newer version 3.4.x which created way more entries than before. Then you have connection server, my friend uses older version, I was able to find later image, but again, no clear advice whether it will work, or whether it is required to update it after upgrading to 3.4.19.
I am also curious if one can just easily replace CU3-01M with CU3-08M that would be able to support HA. They also sell separately so called Inels Bridge, which I believe is just some linux OS with potentially dockered version of HA and MQTT. But - does one still need connection server? What for? There is also Inels Cloud available, says it is supported with certain FW versions, but again, unable to find which are those...But I am already OT here.

**Comment by ds141414** [comment]
Hi!
You beat me to it.  I was on it yesterday as well into the night.  I too installed the latest version of 3.4.19 and managed to get MQTT working.  I can see the topics created in the MQTT Server.  Then there was the mystery on how to make them appear in HA.  It was not clear as you said in the documentation how things are set up for CU3-01M.  After jumping around for a bit I find this https://github.com/epdevlab/elkoep-hacs  After installing that all the bit entities appeared in HA.  So far good news.  

What started happening though was everytime I would push an update to the CU it would ask for which settings to use, project settings or CU settings.  I would choose CU settings and the CU would reboot install the project but I was not able to connect to it after that.  At some point it failed completely and nothing would come on and I had to leave the house for a few hours.  So no switchable lights until I got home, factory reset the unit, upgrade the firmware and then re upload the project.  Up to this point everything is working, CU, CS and MQTT integration.  I will leave it for a few days to see how stable it is.  What I noticed though is that after every CU restart CS would stop reconnecting  to CU and after a reboot it would function OK.

To answer your question regarding  replacing CU, yes it is possible.  They have a guide how to do it.  I will find it and post it here.  If I had to choose I would just upgrade to CU3-08 and would not consider the Bridge as it is too expensive.  The bridge is just Home assistant with the Connection server driver baked in plus all the bells and whistles HA to integrate with other systems.  

I think I am having better success because at some point I did the whole house on a newer version of IDM and I did not have to go through all the conversion process.

**Comment by stomko11** [comment]
Thanks for sharing.
Out of curiosity - which version of CS do you use?
Also - one of interesting things is - it is inconsistent for me how export file is available under CU/immfiles/export.imm. Right now I am getting "404 - Not Found" error. I got same last week after hassle with attempted upgrade and rollback. But few days later, it worked again.

Also on that guide for CU replacement (upgrade) - this is exactly what I had in mind. Not very transparent about sharing these, I found image of CS one one of lolal branches, I believe it was UK. But nothing on local site. But it is all hit and miss.

**Comment by ds141414** [comment]
Hi!  I hit that documentation wall from day one with Inels.  The hardware is good but it has been a nightmare with documentation.  I remember the first time I had to factory reset the unit.  I coud not find a guide anywhere.  They kind of assume that anyone is supposed to know these things.  I am guessing they are more targeted towards installers and integrators rather than us home tinkerers :) 

Anyway back to this.  Have you checked which services are running after the system boots up? When you cannot download the file try telnet to port 80.  That might give you an indication if the http server is running or not.  I would also try to download the file from IDM

My connection server verson is: ver. connection-server-3.362.  If you image for CS is newer than this please share.  I could try it out.

**Comment by stomko11** [comment]
The image I found is 3.362 too. My friend used 3.265. That could somehow contribute to the issue. But still, I would expect that after rolling back to original version, it would work.
The export on CU is not present. CU listens on port 80 as the message thrown is comming from CU's server.

**Comment by ds141414** [comment]
This might be useful, I found it in C:\Users\**<Username>**\ELKO EP\iNELS BUS System\iNELS3 03.04.19 Designer & Manager\Documentation

[Recommended minimal firmware version.txt](https://github.com/user-attachments/files/17445975/Recommended.minimal.firmware.version.txt)

**Comment by stomko11** [comment]
Thanks. I think that next steps here will be just to upgrade to CU3-08M and do everything natively.

**Comment by ds141414** [comment]
Yes, thats what I am thinking in the future.  There is no sense for inels bridge if you already have a running Home Assistant instance.  
Do you have a local supplier for inels hardware or do you order online?

**Comment by stomko11** [comment]
My friend reached out to one of local inels representatives (technical support) just to cofirm the desired plan. As per his instructions, project will need to be reworked. He did not mention any easy way to migrate it over to CU3-08M. But OK, few hours spent is still worth it.
Normally you can't buy anything online here, as it is all sold via partners. But he also offered to sell CU - the price is around 150 EUR for CU3-08M unit. All other units that he has installed (SA3-012M, SA3-04M/06M, DA3-22M, SA3-04M, SA3-012M, WSB3-20/40, IM3-140M) are compatible both with CU3-08M as well as are fully supported by MQTT.

**Comment by ds141414** [comment]
It is the same here in Cyprus, only through local representative.   There are a couple local representatives and since we are such a small place, prices are quite high.  Last time I checked they wanted around 250 euro for the CU3-08M.  For 150 euro I would buy one now :) Do you think your dealer could send one here ?

**Comment by stomko11** [comment]
If he can't, perhaps I could resend it. Would you be interested?

**Comment by ds141414** [comment]
Hi!  Yes definately!

**Comment by stomko11** [comment]
Can we connect somewhere else (email, ...) so that we are not continuing this OT discussion here?

**Comment by ds141414** [comment]
Yes, I was thinking about that too...we went on a huge tangent email me at ds1414@outlook.com

**Comment by KalKal1976** [comment]
Hi only short info from ELKO, there is no compatibility between actors and components used with CU3-01/02 and CU3-08.
But when I see your communication maybe you will find something. :) 
Where are you continue with this discusion?

**Comment by ds141414** [comment]
Hi! What exactly are you trying to achieve? In our case we settled on the conclusion that the best way to control Inels devices was to replace Cu3-01m with CU3-08M.

Now trying to find a way to have both units work on the same Bus

**Comment by KalKal1976** [comment]
I had disucion with some people from Elko (one of them is iNels developer) and they told to me that is not possible run CU3-8 on same instalation as CU3-1/2 (I have version 2). I have question if you have it yet in your instalation and trying run it or no. It will be really interesting info for me.

**Comment by KalKal1976** [comment]
Have you any better place for this discusion for example Discord?

**Comment by ds141414** [comment]
Hi,

Just set up an inels server on discord


https://discord.gg/Kx9zcacs



Sent from my Galaxy



-------- Original message --------
From: KalKal1976 ***@***.***>
Date: 21/12/2024 22:54 (GMT+02:00)
To: JH-Soft-Technology/InelsForHass ***@***.***>
Cc: ds141414 ***@***.***>, Mention ***@***.***>
Subject: Re: [JH-Soft-Technology/InelsForHass] Sensors data availability ? (Issue #15)


Have you any better place for this discusion for example Discord?
